File: q1.py
import sys
from collections import defaultdict

MOD = 1000000007

# Function to perform depth-first search and count good sequences
def dfs(node, parent, color, k, dp):
    if dp[node][color] is not None:
        return dp[node][color]
    
    total_sequences = pow(len(tree), k, MOD)
    invalid_sequences = 1

    for neighbor, edge_color in tree[node]:
        if neighbor != parent:
            sub_result = dfs(neighbor, node, edge_color, k - 1, dp)
            invalid_sequences = (invalid_sequences * sub_result) % MOD

    good_sequences = (total_sequences - invalid_sequences + MOD) % MOD
    dp[node][color] = good_sequences
    return good_sequences

# Input handling
def main():
    input = sys.stdin.read
    data = input().split()
    
    idx = 0
    n = int(data[idx])
    k = int(data[idx + 1])
    idx += 2

    global tree
    tree = defaultdict(list)
    
    for _ in range(n - 1):
        u = int(data[idx])
        v = int(data[idx + 1])
        x = int(data[idx + 2])
        tree[u].append((v, x))
        tree[v].append((u, x))
        idx += 3

    # Initialize dp table
    dp = [[None for _ in range(2)] for _ in range(n + 1)]

    total_good_sequences = 0
    for node in range(1, n + 1):
        total_good_sequences = (total_good_sequences + dfs(node, -1, 0, k, dp)) % MOD

    print(total_good_sequences)

if __name__ == "__main__":
    main()

File: q1.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class q1 {
    private static final int MOD = 1000000007;
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int k = scanner.nextInt();
        
        List<List<Edge>> tree = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            tree.add(new ArrayList<>());
        }
        
        // Read the tree edges and colors
        for (int i = 1; i < n; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            int x = scanner.nextInt();
            tree.get(u).add(new Edge(v, x));
            tree.get(v).add(new Edge(u, x));
        }
        
        // Initialize dp array
        long[][] dp = new long[n + 1][k + 1];
        
        // DFS to fill dp array
        dfs(1, -1, tree, dp, k);
        
        // Calculate total number of good sequences
        long totalGoodSequences = 0;
        for (int v = 1; v <= n; v++) {
            totalGoodSequences = (totalGoodSequences + dp[v][k]) % MOD;
        }
        
        // Print the result
        System.out.println(totalGoodSequences);
    }
    
    private static void dfs(int v, int parent, List<List<Edge>> tree, long[][] dp, int k) {
        dp[v][0] = 1; // Base case: one way to have a sequence of length 0 (empty sequence)
        
        for (Edge edge : tree.get(v)) {
            int u = edge.to;
            if (u == parent) continue; // Skip the parent node
            
            dfs(u, v, tree, dp, k);
            
            // Update dp[v][j] for j = 1 to k
            for (int j = 1; j <= k; j++) {
                dp[v][j] = (dp[v][j] + dp[u][j - 1]) % MOD;
                if (edge.color == 1) {
                    dp[v][j] = (dp[v][j] + dp[u][j - 1]) % MOD;
                }
            }
        }
    }
    
    static class Edge {
        int to;
        int color;
        
        Edge(int to, int color) {
            this.to = to;
            this.color = color;
        }
    }
}
